rasa/core/featurizers/single_state_featurizer.py

```diff
@@ -40,127 +102,227 @@ class SingleStateFeaturizer:
-        #  entities and the concatenated entity and roles/groups. We do not
-        #  distinguish between entities and roles/groups right now.
-        #  we return a list to anticipate that
-        return [
-            EntityTagSpec(
-                tag_name=ENTITY_ATTRIBUTE_TYPE,
-                tags_to_ids=tag_id_index_mapping,
-                ids_to_tags={value: key for key, value in tag_id_index_mapping.items()},
-                num_tags=len(tag_id_index_mapping),
-            )
-        ]
+            # nlu pipeline didn't create features for user or action
+            # this might happen, for example, when we have action_name in the state
+            # but it did not get featurized because only character level
+            # CountVectorsFeaturizer was included in the config.
+            name_attribute = get_name_attribute(sub_state)
+
+            if substate_type == PREVIOUS_ACTION:
+                substate_features = self.featurize_substate_via_interpreter(
+                    sub_state,
+                    interpreter,
+                    exclude={ENTITIES},
+                    keep_only_sparse_seq_converted_to_sent={INTENT, ACTION_NAME},
+                    fallback_to_custom_encoding={name_attribute},
+                    fallback_is_sparse=True,
+                )
+
+            elif substate_type == USER and is_prev_action_listen_in_state(state):
+                # featurize user only if it is "real" user input,
+                # i.e. input from a turn after action_listen
+                substate_features = self._featurize_substate_via_interpreter(
+                    sub_state,
+                    interpreter,
+                    exclude={ENTITIES},
+                    keep_only_sparse_seq_converted_to_sent={INTENT, ACTION_NAME},
+                    fallback_to_custom_encoding={name_attribute},
+                    fallback_is_sparse=True,
+                )
+                if sub_state.get(ENTITIES):
+                    state_features[
+                        ENTITIES
+                    ] = self._featurize_attribute_via_custom_encoding(
+                        sub_state=sub_state, attribute=ENTITIES,
+                    )
+
+            elif substate_type in {SLOTS, ACTIVE_LOOP}:
+                substate_features = {
+                    substate_type: self._featurize_attribute_via_custom_encoding(
+                        sub_state=sub_state, attribute=substate_type,
+                    )
+                }
+            else:
+                substate_features = {}
+
+            state_features.update(substate_features)
+
+        return state_features
+
+    def featurize_action(
+        self, action: Text, interpreter: NaturalLanguageInterpreter
+    ) -> Dict[Text, List["Features"]]:
+        """
+
+        """
+        # FIXME: if prepare_for_training has not been called then action_texts will
+        # be an empty list and this always defaults to "ACTION_NAME" - is this intended?
```

Comment:
We need to get rid of the `prepare_for_training` and put it into the `__init__` somehow.

Reply:
agree!